### ISSUE TITLE
Ruff fails when a directory ends with .py

### ISSUE THREAD

**Issue by konstin** [astral-sh/ruff#5739]
Ruff gets confused when a directory is named like a python file (ends with `.py`):

```shell
$ mkdir repro
$ mkdir repro/some_folder.py
$ cargo run --bin ruff -- --no-cache repro
error: Failed to lint repro/some_folder.py: Is a directory (os error 21)
  Caused by: Is a directory (os error 21)
repro/some_folder.py:1:1: E902 Is a directory (os error 21)
  Caused by: Is a directory (os error 21)
Found 1 error.
```

**Comment by zanieb** [comment]
@konstin is this a good first issue?

**Comment by charliermarsh** [comment]
(From my perspective, seems like a good issue, will probably require digging into how we collect the "files to lint".)

**Comment by konstin** [comment]
i've confirmed it's not related to #5611, so yes it's a good first issue

**Comment by harupy** [comment]
can I work on this?

**Comment by harupy** [comment]
Is this the part that needs to be fixed?

https://github.com/astral-sh/ruff/blob/f029f8b7840e4477f2f42d8f589c4ac290a249e6/crates/ruff/src/resolver.rs#L332-L350

**Comment by harupy** [comment]
Testing this change:

```diff
diff --git a/crates/ruff/src/resolver.rs b/crates/ruff/src/resolver.rs
index 0ad0fa244..0d851135c 100644
--- a/crates/ruff/src/resolver.rs
+++ b/crates/ruff/src/resolver.rs
@@ -330,9 +330,12 @@ pub fn python_files_in_path(
             }
 
             if result.as_ref().map_or(true, |entry| {
+                if !(entry.file_type().map_or(false, |ft| ft.is_file())) {
+                    return false;
+                }
                 if entry.depth() == 0 {
                     // Accept all files that are passed-in directly.
-                    entry.file_type().map_or(false, |ft| ft.is_file())
+                    true
                 } else {
                     // Otherwise, check if the file is included.
                     let path = entry.path();
```

Result:

```
> tree dir
dir
├── dir.py
│   └── file.py
└── file.py

1 directory, 2 files

> cargo run -p ruff_cli -- check dir --no-cache
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/ruff check dir --no-cache`
dir/dir.py/file.py:1:7: F821 Undefined name `a`
dir/file.py:1:7: F821 Undefined name `a`
Found 2 errors.
```
